torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,134 @@ namespace rpc {
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(true);
+  }
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(false);
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+}
 
-RpcAgent::~RpcAgent() = default;
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+  auto future = std::make_shared<FutureMessage>();
+  auto fm = send(to, std::move(message));
+  fm->addCallback([future](
+                      const rpc::Message& message,
+                      const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      LOG(INFO) << "Send try 1 failed\n";
+    } else {
+      // First try succeeded and we can mark the original future as completed.
+      future->markCompleted(message);
+    }
+  });
+  steady_clock_time_point newTime =
+      std::chrono::time_point_cast<std::chrono::milliseconds>(
+          std::chrono::steady_clock::now() + kRpcRetryDuration);
+  rpcRetryMap_.emplace(
+      std::piecewise_construct,
+      std::forward_as_tuple(newTime),
+      std::forward_as_tuple(RpcRetryInfo(
+          to, std::move(message), fm, future, /* retryCount */ 1, maxRetries)));
+  rpcRetryMapCV_.notify_one();
+
+  return future;
+}
+
+void RpcAgent::retryExpiredRpcs() {
+  while (rpcAgentRunning_.load()) {
+    // TODO(osalpekar): This is temporary solution for handling empty map and
+    // sleeping until RPC retry duration is expired per Milestone 1.
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
+    rpcRetryMapCV_.wait(lock, [&, this] {
+      return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+    });
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+    steady_clock_time_point earliestTimeout = rpcRetryMap_.begin()->first;
+    if (std::chrono::steady_clock::now() < earliestTimeout) {
+      std::this_thread::sleep_until(earliestTimeout);
+    }
+    RpcRetryInfo& earliestRpc = rpcRetryMap_.begin()->second;
+
+    // If the future for this RPC's last try was not completed, we will re-try
+    // sending the RPC and attach a callback to the returned future that marks
+    // the original future (which was returned to the user) as completed if the
+    // retry succeeds. We also compute a new time in the future (based on
+    // exponential backoff) at which this RPC would be re-tried next and add it
+    // to the rpcRetryMap_.
+    if (!earliestRpc.retryFuture_->completed()) {
+      auto fm = send(earliestRpc.to_, std::move(earliestRpc.message_));
```

Comment:
This `fm` also lives in `futures_` in `ProcessGroupAgent`? Which will never get cleared if the the ack didn't come back?

Reply:
The un-ACKed RPC’s in the `futures_` map are being cleared if they aren't responded to after `rpcTimeout` by the `futureTimeoutThread_` in ProcessGroupAgent